libraries/plugins/witness/witness.cpp

```diff
@@ -250,7 +253,7 @@ block_production_condition::block_production_condition_enum witness_plugin::mayb
    }
 
    fc::time_point_sec scheduled_time = db.get_slot_time( slot );
-   graphene::chain::public_key_type scheduled_key = scheduled_witness( db ).signing_key;
+   graphene::chain::public_key_type scheduled_key = *db.find_witness_key_from_cache( scheduled_witness ); // should be valid
```

Comment:
I think instead of using the cache, witness_plugin could simply connect to the ``database::applied_block`` signal and use that to retrieve the latest keys of the witness(es) it is interested in.

Reply:
Essentially you mean to move the cache to witness plugin. Because there is no signal when a block is popped out, in order to deal with chain reorganization, the cache should be in (extended) chain state. Then, in order to make it work, need to either replay the whole chain to enable the plugin, or add code to refresh the cache (copy data from main chain state to the extended part) on startup. Then, in case when there is a reorganization to the startup head block, need to make sure the cache don't get emptied.

I didn't see it's simpler so far.

Reply:
I agree that ideally it's better to put the logic in witness plugin because block generation is actually witness plugin's business.

> Blocks are only popped when switching forks, which means there will be block pushes immediately after the pop, which means the cache will automatically be updated in that case.

One scenario to consider: if put the cache in witness plugin, if we code the plugin to track applied operations only, after a block is popped out, the cache may become inconsistent if the popped out block contains a `witness_update_operation`, when pushing a new block, if there is no the same `witness_update_operation` in the new block, the cache would still be inconsistent. To solve this, we need additional code to detect if a chain reorganization has occurred and refresh the cache accordingly, if not simply refresh the whole cache on every block. However, when refreshing the cache, we need to fetch data from main chain state, since signalling is a multi-threading thing, the main state could have changed when fetching data, which means the cache is not 100% reliable.

By putting the cache in database class we can guarantee its consistency. Perhaps there will still be edge cases, for example, witness plugin tries to generate a block in the middle when database is pushing a block.

Another thing is we don't have unit test for witness plugin, so adding more code there means more work / risks.

Reply:
The witness plugin *typically* needs to keep track of the keys of only one witness. Refreshing this after each pushed block should have minimal cost, and only for the node in question.

Reply:
I think the applied_block signal is processed in sync right after the block has been applied. There shouldn't be any tx's pushed in between.
https://github.com/bitshares/bitshares-core/blob/2.0.180612/libraries/chain/db_block.cpp#L550

Reply:
Oh, right, I forgot that. The state should be clean when processing applied_block event.

By the way, why we need to sleep here and there in test cases? IIRC sometimes chain_test will segfault without sleeping. E.G. https://github.com/bitshares/bitshares-core/blob/95906d8100198a070ab1894e465a8611bc559e56/tests/tests/market_rounding_tests.cpp#L76

Reply:
I think it's required in some places where API callbacks happen. These are asynchronous.
Shouldn't be necessary in market_rounding_tests IMO.